pkgs/development/compilers/llvm/14/mlir/default.nix

```diff
@@ -0,0 +1,88 @@
+{ lib, stdenv, llvm_meta
+, monorepoSrc, runCommand
+, cmake
+, libllvm
+, version
+
+# TODO: python bindings
+# TODO: ROCm, CUDA ?
+# TODO: Tests (unit+lit), not integration
+, enableRunners ? stdenv.hostPlatform == stdenv.buildPlatform
+, enableVulkan ? enableRunners
+, vulkan-headers
+, vulkan-loader
+}:
+
+let
+  # List of runners to build, native-only
+  runners = lib.optionals enableRunners ([
+    "cpu-runner" "spirv-cpu-runner"
+  ] ++ lib.optional enableVulkan "vulkan-runner");
+  # List of binaries to build + install
+  # LLVM_{BUILD,INSTALL}_UTILS=ON doesn't seem to work
+  bins = map (n: "mlir-" + n) (runners ++ [
+    "linalg-ods-yaml-gen" "tblgen" # needed for cross
+    "lsp-server" "opt" "pdll" "reduce" "translate" # misc utilities
+  ]);
+in
+stdenv.mkDerivation rec {
+  pname = "mlir";
+  inherit version;
+
+  src = runCommand "${pname}-src-${version}" {} ''
+    mkdir -p "$out"
+    cp -r ${monorepoSrc}/cmake "$out"
+    cp -r ${monorepoSrc}/${pname} "$out"
+  '';
+
+  sourceRoot = "${src.name}/${pname}";
+
+  patches = [
+    ./gnu-install-dirs.patch
+  ];
+
+  outputs = [ "out" "lib" "dev" ];
+
+  nativeBuildInputs = [ cmake ];
+  buildInputs = [ libllvm ]
+    ++ lib.optionals enableVulkan [ vulkan-headers vulkan-loader ];
+
+  cmakeFlags = [
+    # Documentation suggests packagers may wish to disable, do so until needed
+    "-DMLIR_INSTALL_AGGREGATE_OBJECTS=OFF"
+  ] ++ lib.optionals enableRunners ([
+    "-DMLIR_ENABLE_SPIRV_CPU_RUNNER=ON"
+  ] ++ lib.optional enableVulkan "-DMLIR_ENABLE_VULKAN_RUNNER=ON");
+
+  # Patch around check for being built native (maybe because not built w/LLVM?)
+  postPatch = lib.optionalString enableRunners ''
+    for x in **/CMakeLists.txt; do
+      substituteInPlace "$x" --replace 'if(TARGET ''${LLVM_NATIVE_ARCH})' 'if (1)'
```

Comment:
In LLVM git / v15 there is inverse check too, so we'd need to add this to avoid build breakage:

```
substituteInPlace "$x" --replace 'if(NOT TARGET ''${LLVM_NATIVE_ARCH})' 'if (0)'
```

Reply:
globstar doesn't seem to be enabled in stdenv/bash, so the for loop doesn't process files recursively. We can use `find -name CMakeLists.txt | while read f; do ...; done` instead.